### PR TITLE
feat(auto-salvage): v2 support

### PR DESCRIPTION
**Pull request by c3y1huang**
#### Which issue(s) this PR fixes:

Issue longhorn/longhorn#8430

#### What this PR does / why we need it:

Skip creating lvol bdev for a new replica instance if it exists.

#### Special notes for your reviewer:

`None`

#### Additional documentation or context

`None`

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -4,6 +4,10 @@ go 1.22.0
 
 toolchain go1.23.1
 
+replace github.com/longhorn/types => github.com/c3y1huang/types v0.0.0-20240927004343-3ad221f0cbfc
+
+replace github.com/longhorn/go-common-libs => github.com/c3y1huang/go-common-libs v0.0.0-20240927004544-ee33d7bb917c
```
Comment (c3y1huang) [inline go.mod:9]: TODO: update when dependency PRs are merged.

File: pkg/spdk/engine.go [hunk 2]
```diff
@@ -271,6 +280,76 @@ func (e *Engine) Create(spdkClient *spdkclient.Client, replicaAddressMap map[str
...
+	largestVolumeHeadSize := volumeHeadSizeSorted[len(volumeHeadSizeSorted)-1]
+	e.log.Infof("Filtering salvage candidates by volume head size (%v) from %+v", largestVolumeHeadSize, volumeHeadSizeToReplicaNames)
+
+	salvageCandidates := map[string]bool{}
+	for _, replicaName := range volumeHeadSizeToReplicaNames[largestVolumeHeadSize] {
+		salvageCandidates[replicaName] = true
+	}
+
+	for replicaName := range e.ReplicaAddressMap {
+		if _, exist := salvageCandidates[replicaName]; !exist {
```
Comment (DamiaSan) [inline pkg/spdk/engine.go:341]: How can `replicaName` not exist in `salvageCandidates`? This map has been populated by `volumeHeadSizeToReplicaNames`, which has been populated by `ReplicaAddressMap`.

File: pkg/spdk/engine.go [hunk 3]
```diff
@@ -206,7 +207,15 @@ func (e *Engine) Create(spdkClient *spdkclient.Client, replicaAddressMap map[str
 		e.ReplicaAddressMap = replicaAddressMap
 		e.log = e.log.WithField("replicaAddressMap", replicaAddressMap)
 
-		e.checkAndUpdateInfoFromReplicaNoLock()
+		if salvageRequested {
```
Comment (shuo-wu) [inline pkg/spdk/engine.go:210]: Does the engine need to try to connect all replicas before a salvage?
If NO, then the mode check inside `updateInfoToFilterReplicaSalvageCandidates` makes no sense.

File: pkg/spdk/engine.go (hunk 2 above)
Comment (c3y1huang) [reply pkg/spdk/engine.go]: Fixed, thank you!

File: pkg/spdk/engine.go (hunk 3 above)
Comment (c3y1huang) [reply pkg/spdk/engine.go]: Fixed. Thank you!

**Review by DamiaSan — APPROVED** [review]

File: go.mod [hunk 4]
```diff
@@ -4,11 +4,13 @@ go 1.22.0
 
 toolchain go1.23.2
 
+replace github.com/longhorn/types => github.com/c3y1huang/types v0.0.0-20241027234820-01e88dd78e41
```
Comment (c3y1huang) [inline go.mod:7]: TODO: update when dependency PRs are merged.

File: pkg/spdk/engine.go [hunk 5]
```diff
@@ -271,6 +285,72 @@ func (e *Engine) Create(spdkClient *spdkclient.Client, replicaAddressMap map[str
 	return e.getWithoutLock(), nil
 }
 
+// updateReplicaModeToFilterSalvageCandidates updates e.ReplicaModeMap and filters
+// out the replicas that are not eligible for salvage based on the volume head size.
+//
+// It iterates through all replicas and:
+//   - Skips the replicas with mode ERR.
+//   - Gets the volume head size of each replica.
+//   - Selects replicas with the largest volume head size as salvage candidates.
```
Comment (derekbit) [inline pkg/spdk/engine.go:294]: Let's see if there is any better method to select candidates in the future.

cc @DamiaSan @shuo-wu

**Review by derekbit — DISMISSED** [review]
LGTM

**Comment by derekbit** [comment]
@c3y1huang We can merge it after resolving the conflict. Thank you.

**Comment by c3y1huang** [comment]
The logic in this PR needs to be updated to incorporate the changes introduced in https://github.com/longhorn/longhorn-spdk-engine/pull/208.

**Comment by c3y1huang** [comment]
@coderabbitai review

**Review by derekbit — APPROVED** [review]

**Pull request merged**